src/gleam/list.gleam

```diff
@@ -790,7 +790,19 @@ pub fn find_map(
 pub fn all(in list: List(a), satisfying predicate: fn(a) -> Bool) -> Bool {
   case list {
     [] -> True
-    [x, ..rest] -> predicate(x) && all(rest, predicate)
+    list -> all_tail_recursive(list, predicate, True)
+  }
+}
+
+fn all_tail_recursive(
+  list: List(a),
+  predicate: fn(a) -> Bool,
+  accumulator: Bool,
+) -> Bool {
+  case list {
+    [] -> accumulator
+    [x, ..rest] ->
+      all_tail_recursive(rest, predicate, accumulator && predicate(x))
```

Comment:
This continues iterating even if the predicate function has returned `False`.

Let's have a test for the short circuiting, something like this:
```gleam
[1, 2, 3]
|> list.all(fn(x) {
  case x {
    1 -> True
    2 -> False
    // Crash to ensure we short-circuit
    _ -> assert True = False
  }
})